File: Euclid/Game/Room/Mapping/RoomMapping.cs
namespace Euclid.Game
{
    public class RoomMapping : ILoadable
    {
        #region Fields

        private Room room;
        private RoomModel model;

        #endregion

        #region Properties

        public RoomTile[,] Tiles { get; private set; }

        #endregion

        #region Constructors

        public RoomMapping(Room room)
        {
            this.room = room;
            this.model = room.Model;
        }

        #endregion

        #region Public methods

        /// <summary>
        /// Regenerate the room map
        /// </summary>
        public void Load()
        {
            Tiles = new RoomTile[model.MapSizeX, model.MapSizeY];

            for (int y = 0; y < model.MapSizeY; y++)
            {
                for (int x = 0; x < model.MapSizeX; x++)
                {
                    Tiles[x, y] = new RoomTile(
                        room,
                        new Position(x, y),
                        model.TileHeights[x, y],
                        model.TileStates[x, y]
                    );
                }
            }

            foreach (var item in room.EntityManager.GetEntities<Item>())
            {
                if (item.Definition.Data.IsWallItem)
                    continue;

                foreach (Position position in AffectedTile.GetAffectedTiles(item))
                {
                    var tile = position.GetTile(room);

                    if (tile == null)
                        continue;

                    tile.AddItem(item);
                }
            }

        }

        /// <summary>
        /// Mapping item handler for adding/removing item to collision map
        /// </summary>
        public void AddItem(Item item)
        {
            // Add item to tile
            foreach (var affectedPosition in AffectedTile.GetAffectedTiles(item))
            {
                var roomTile = affectedPosition.GetTile(room);

                if (roomTile == null)
                    continue;

                roomTile.AddItem(item);
            }

        }

        /// <summary>
        /// Mapping item handler for adding/removing item to collision map
        /// </summary>
        public void RemoveItem(Item item)
        {
            // Remove item from tile
            foreach (var affectedPosition in AffectedTile.GetAffectedTiles(item))
            {
                var roomTile = affectedPosition.GetTile(room);

                if (roomTile == null)
                    continue;

                roomTile.RemoveItem(item);
            }

        }

        #endregion
    }
}

File: euclid/src/main/java/euclid/game/room/RoomMapping.java
package euclid.game.room;

import euclid.game.ILoadable;
import euclid.game.item.Item;
import euclid.game.pathfinder.AffectedTile;
import euclid.game.pathfinder.Position;

public class RoomMapping implements ILoadable {

    private final Room room;
    private final RoomModel model;
    private RoomTile[][] tiles;

    public RoomMapping(Room room) {
        this.room = room;
        this.model = room.getModel();
    }

    @Override
    public void load() {
        tiles = new RoomTile[model.getMapSizeX()][model.getMapSizeY()];

        for (int y = 0; y < model.getMapSizeY(); y++) {
            for (int x = 0; x < model.getMapSizeX(); x++) {
                tiles[x][y] = new RoomTile(
                        room,
                        new Position(x, y),
                        model.getTileHeights()[x][y],
                        model.getTileStates()[x][y]
                );
            }
        }

        for (Item item : room.getEntityManager().getEntities(Item.class)) {
            if (item.getDefinition().getData().isWallItem())
                continue;

            for (Position position : AffectedTile.getAffectedTiles(item)) {
                RoomTile tile = position.getTile(room);
                if (tile == null)
                    continue;
                tile.addItem(item);
            }
        }
    }

    public void addItem(Item item) {
        for (Position affectedPosition : AffectedTile.getAffectedTiles(item)) {
            RoomTile roomTile = affectedPosition.getTile(room);
            if (roomTile == null)
                continue;
            roomTile.addItem(item);
        }
    }

    public void removeItem(Item item) {
        for (Position affectedPosition : AffectedTile.getAffectedTiles(item)) {
            RoomTile roomTile = affectedPosition.getTile(room);
            if (roomTile == null)
                continue;
            roomTile.removeItem(item);
        }
    }

    public RoomTile[][] getTiles() {
        return tiles;
    }
}
